Countdown: Continue PR #3485

rasikapohankar opened the pull request:
<!-- Use the appropriate format for your Pull Request title above ^^^^^:

If this is a bug fix:
{IA Name}: {Description of change}

If this is a New Instant Answer:
New {IA Name} Spice

If this is something else:
{Tests/Docs/Other}: {Short Description}

-->

## Description of new Instant Answer, or changes
<!-- What does this new Instant Answer do? What changes does this PR introduce? -->
Implements changed design, fixes #3500

Design reference:
![countdown v2 design](https://cloud.githubusercontent.com/assets/8025692/23174733/51957562-f883-11e6-82ff-125d8bdf4909.jpg)

## Related Issues and Discussions
<!-- Link related issues here to automatically close them when PR is merged -->
<!-- E.g. "Fixes #1234" -->
#3500 

## People to notify
<!-- Please @mention any relevant people/organizations here: -->
@moollaza 
<!-- DO NOT REMOVE -->
---

<!-- The Instant Answer ID can be found by clicking the `?` icon beside the Instant Answer result on DuckDuckGo.com -->
Instant Answer Page: https://duck.co/ia/view/countdown
<!-- FILL THIS IN:                           ^^^^ -->

daxtheduck commented:
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

mintsoft commented on lib/DDG/Goodie/Countdown.pm:
```diff
@@ -13,10 +13,12 @@ zci answer_type => 'countdown';
 
 zci is_cached => 1;
 
+my $goodieVersion = $DDG::GoodieBundle::OpenSourceDuckDuckGo::VERSION // 999;
```
?

rasikapohankar replied:
@mintsoft It is used to form the path of the alarm sound in production ( #3500 )

mintsoft commented on t/Countdown.t:
```diff
@@ -34,9 +40,9 @@ set_fixed_time("2016-04-15T15:31:02Z");
-    'time until 1st June 2016'            => build_test("1st June 2016", 4019338 , "01 Jun 2016 00:00:00 EDT"),
-    'how long until 31st December 2016'   => build_test("31st December 2016", 22426138, "31 Dec 2016 00:00:00 EST"),
-    'countdown to tomorrow'               => build_test("tomorrow", 86400, "16 Apr 2016 11:31:02 EDT"),
+    'time until 1st June 2016'            => build_test("1st June 2016", 4019338 , "June 01, 2016, 12:00:00 AM"),
```
Usually when we test something that uses the DateRole we tend to use ISO8601 for the date format; there's no reason to test the Role here, as it's already tested independently

rasikapohankar replied:
@mintsoft I'm unclear about this. It checks the difference between 2016-04-15T15:31:02Z and 1st June 2016 for the countdown. It's not testing the Role..

mintsoft replied:
The Dates Role parses "1st june 2016", so normally we'd use like `time until 2016-04-15` as the test case

rasikapohankar replied:
@mintsoft I understand now! Updated the test file :+1:

mintsoft commented on share/goodie/countdown/countdown.js:
```diff
@@ -106,48 +72,52 @@ DDH.countdown = DDH.countdown || {};
                         duration = getCountdown(duration);
+                        item.set({ year: duration.years(), month: duration.months(), day: duration.days(), hour: duration.hours(), minute: duration.minutes(), second: duration.seconds() });
+                        $(".zci--countdown").find('.c-base__sub').addClass("tx-clr--grey");
                     }, 1000);
```
With an interval of 1000 we might see skips of 2 seconds as the timer firing isn't going to be exact, we can flip this out for a shorter interval if it's something you're concerned about

rasikapohankar replied:
If we make the interval shorter, the countdown will run faster as this renders the template after every second :sweat_smile: I don't think the interval should be changed. Do you think this will be unreliable?

mintsoft replied:
I don't think it's a major concern, we'll see how it looks. If we see a 2 second jump then it'll need decreasing, but it's probably fine

mintsoft commented on share/goodie/countdown/countdown.js:
```diff
@@ -5,46 +5,13 @@ DDH.countdown = DDH.countdown || {};
         cachedPlayer, soundIsPlaying = false,
         SOUND_NAME = "alarm-sound",
-        soundUrl = 'share/goodie/countdown/alarm.mp3',
+        soundUrl,  // 'share/goodie/countdown/goodie-repo-version/alarm.mp3',
```
Might as well get shot of the comment as it's defined below

mintsoft commented on t/Countdown.t:
```diff
@@ -11,18 +11,24 @@ use DDG::Test::Location;
 zci answer_type => "countdown";
 zci is_cached   => 1;
 
+my $goodieVersion = $DDG::GoodieBundle::OpenSourceDuckDuckGo::VERSION // 999;
```
Again, I'm not sure why this is here/needed ?

rasikapohankar replied:
@mintsoft I had added this to include the `goodie_version` returned from Countdown.pm..

mintsoft commented on t/Countdown.t:
```diff
@@ -11,18 +11,24 @@ use DDG::Test::Location;
+                goodie_version => $goodieVersion
             },
             templates => {
                 group => "text",
```
If you're using a custom template like this `group => 'base'` is preferred

moollaza replied:
`text` is actually necessary here because we're defining the `subtitle` which is handled by the template. If we switch to base we'd need to add more markup for that to our template 👍

rasikapohankar commented:
@mintsoft @moollaza Can this be deployed to beta?

moollaza commented:
Deploying to beta 👍

moollaza commented on share/goodie/countdown/countdown.js:
```diff
@@ -106,48 +72,52 @@ DDH.countdown = DDH.countdown || {};
                     setInterval(function() {
                         duration = getCountdown(duration);
+                        item.set({ year: duration.years(), month: duration.months(), day: duration.days(), hour: duration.hours(), minute: duration.minutes(), second: duration.seconds() });
+                        $(".zci--countdown").find('.c-base__sub').addClass("tx-clr--grey");
```
@rasikapohankar we should use a CSS rule instead of repeatedly applying a class every second

rasikapohankar replied:
@moollaza This was to apply the class to the `subtitle`. I'll include the subtitle in the handlebars and apply to it the same class as the already defined `subtitle` (`.c-base__sub`). It is working on duckpan. Just wanted to confirm that it will not break anything else?

moollaza commented:
@rasikapohankar aside from my one comment, this LGTM :+1:

moollaza commented:
Oh one bug I just noticed. It seems to not account for the daylight savings shift in my timezone. Can you confirm if this is a moment.js bug?

![countdown_to_april_11th_at_duckduckgo](https://cloud.githubusercontent.com/assets/873785/24892241/130a63d4-1e49-11e7-89b6-f8d2efe31f58.png)

Midnight is 5 minutes away, not 1 hour and 5 minutes.

rasikapohankar commented:
> Can you confirm if this is a moment.js bug?

@moollaza We have only parsed the input string using the Dates Role and found the initial difference between that date and `now`. :slightly_frowning_face:

I had tried to calculate the difference using moment.js at the very start but removed that part later. I will check if it gives the correct answer.

mintsoft commented:
@rasikapohankar @moollaza I think that problem comes from the fact that `$diff` has been calculated by subtracting the `epoch` from each other. IIRC `$dt->delta_ms( $datetime )` should be used for DST aware differences; however it may not have enough information to accurately determine the DST settings for the user. If that's the case, it probably makes sense to use the role to normalise the date into ISO8601 and then pass that through to the frontend, use moment to calculate the diff so that it can handle the timezone based on the browser settings

rasikapohankar commented:
@moollaza @mintsoft I've changed the code to calculate the difference in JS using momentjs and used it to show the countdown, also kept the parsing of the date from query and difference calculation in Perl as it is to avoid going ahead on incorrect input. Please take a look.

@moollaza After using template type option `content`, the subtitle was shown at the top. So removed the class applied to `subtitle` completely, it is now #888.
Also corrected a mistake at the end of the countdown.

@moollaza Can you please deploy to beta again?

moollaza commented:
@rasikapohankar Thanks! This LGTM :+1:

Great work 💯

rasikapohankar commented:
@moollaza  Thanks for merging! I was not aware that this had been deployed to beta. It must have shown the correct answer now?
